### PR TITLE
[MPS] Add repeat_interleave to MPS

### LINKED ISSUE #87219: [MPS] Add support for aten::repeat_interleave for MPS backend
### 🐛 Describe the bug

First time contributors are welcome!
Add support for [aten::repeat_interleave](https://pytorch.org/docs/stable/generated/torch.repeat_interleave.html) for MPS backend. Generic support for adding operations to MPS backend is captured here: https://github.com/pytorch/pytorch/wiki/MPS-Backend#adding-op-for-mps-backend

### Versions

N/A

cc @kulinseth @albanD @malfet @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by jazzysoggy**
Fixes #87219

Implements new ``repeat_interleave`` function into ``aten/src/ATen/native/mps/operations/Repeat.mm``
Adds it to ``aten/src/ATen/native/native_functions.yaml``
Adds new test ``test_repeat_interleave`` to ``test/test_mps/py``

### CODE REVIEW

**Review by kulinseth — APPROVED** [review]
Looks good. @jazzysoggy please fix the lint issue.

File: aten/src/ATen/native/mps/operations/Repeat.mm [hunk 1]
```diff
@@ -179,5 +179,51 @@ Tensor repeat_mps(const Tensor& self, IntArrayRef repeats) {
...
+  } else if (repeats.dim() == 1) {
+    TORCH_CHECK(
+        repeats.size(0) == input.size(dim.value()),
+        "repeats must have the same size as input along dim")
+  } else {
+    AT_ERROR("repeats must be 0-dim or 1-dim tensor");
+  }
+
+  auto output = input.index_select(
+      dim.value(), at::repeat_interleave(repeats_, output_size));
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/Repeat.mm:215]: This won't work. You need to implement this operation. You can write a simple Metal kernel to implement this functionality.

**Review by kulinseth — CHANGES_REQUESTED** [review]
@jazzysoggy please fix the lint issue.

Also what you have is just a copy of Repeat.cpp core file. The backend need to implement the aforementioned interface.

File: aten/src/ATen/native/mps/operations/Repeat.mm (hunk 1 above)
Comment (jazzysoggy) [reply aten/src/ATen/native/mps/operations/Repeat.mm]: Sorry, this one apparently was not fully synced with the written Metal kernel implementation.

**Review by kulinseth — APPROVED** [review]
Please fix the Lint error, otherwise it looks good.

**Comment by kulinseth** [comment]
Also there is Build failure:
```
/Users/runner/work/pytorch/pytorch/aten/src/ATen/native/mps/operations/Repeat.mm:216:105: error: 'NSNumber' may not respond to 'UTF8String' [-Werror]
[3056](https://github.com/pytorch/pytorch/actions/runs/3482646873/jobs/5825210002#step:10:3057)
    string key = "repeat_interleave_mps:" + getTensorsStringKey({repeats})+ ":" + string([repeats_shape UTF8String]);
[3057](https://github.com/pytorch/pytorch/actions/runs/3482646873/jobs/5825210002#step:10:3058)
                                                                                          ~~~~~~~~~~~~~ ^
```

**Comment by jazzysoggy** [comment]
@kulinseth I think I need a little help. Not quite sure how to properly write a native metal kernel. I think I've got the framework correct, but just not accessing the correct namespaces for certain objects or defining functions in the right places.

**Comment by kulinseth** [comment]
> @kulinseth I think I need a little help. Not quite sure how to properly write a native metal kernel. I think I've got the framework correct, but just not accessing the correct namespaces for certain objects or defining functions in the right places.

@jazzysoggy , is this still an issue. Can you please paste the error message you are running into?

**Comment by jazzysoggy** [comment]
> > @kulinseth I think I need a little help. Not quite sure how to properly write a native metal kernel. I think I've got the framework correct, but just not accessing the correct namespaces for certain objects or defining functions in the right places.
> 
> @jazzysoggy , is this still an issue. Can you please paste the error message you are running into?

No, no longer facing this issue anymore. The entire thing compile, just with testing errors now.

‘’FAIL [13.217s]: test_no_warning_on_import (__main__.TestFallbackWarning)
[3541](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3544)
----------------------------------------------------------------------
[3542](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3545)
Traceback (most recent call last):
[3543](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3546)
  File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/test_mps.py", line 7033, in test_no_warning_on_import
[3544](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3547)
    self.assertEqual(out, "")
[3545](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3548)
  File "/Users/ec2-user/runner/_work/_temp/conda_environment_3509465096/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 2555, in assertEqual
[3546](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3549)
    assert_equal(
[3547](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3550)
  File "/Users/ec2-user/runner/_work/_temp/conda_environment_3509465096/lib/python3.9/site-packages/torch/testing/_comparison.py", line 1118, in assert_equal
[3548](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3551)
    raise error_metas[0].to_error(msg)
[3549](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3552)
AssertionError: String comparison failed: '[W OperatorEntry.cpp:159] Warning: Overri[535 chars]l)\n' != ''
[3550](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3553)
- [W OperatorEntry.cpp:159] Warning: Overriding a previously registered kernel for the same operator and the same dispatch key
[3551](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3554)
-   operator: aten::repeat_interleave.Tensor(Tensor repeats, *, int? output_size=None) -> Tensor
[3552](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3555)
-     registered at /Users/runner/work/pytorch/pytorch/build/aten/src/ATen/RegisterSchema.cpp:6
[3553](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3556)
-   dispatch key: MPS
[3554](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3557)
-   previous kernel: registered at /Users/runner/work/pytorch/pytorch/build/aten/src/ATen/RegisterCPU.cpp:30830
[3555](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3558)
-        new kernel: registered at /Users/runner/work/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:48 (function registerKernel)
[3556](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3559)’’

**Comment by jazzysoggy** [comment]
@kulinseth I’m able to compile, but I currently have issues with passing the test. Here’s the error log for the failed test

FAIL [13.217s]: test_no_warning_on_import (__main__.TestFallbackWarning)
[3541](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3544)
----------------------------------------------------------------------
[3542](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3545)
Traceback (most recent call last):
[3543](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3546)
  File "/Users/ec2-user/runner/_work/pytorch/pytorch/test/test_mps.py", line 7033, in test_no_warning_on_import
[3544](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3547)
    self.assertEqual(out, "")
[3545](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3548)
  File "/Users/ec2-user/runner/_work/_temp/conda_environment_3509465096/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py", line 2555, in assertEqual
[3546](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3549)
    assert_equal(
[3547](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3550)
  File "/Users/ec2-user/runner/_work/_temp/conda_environment_3509465096/lib/python3.9/site-packages/torch/testing/_comparison.py", line 1118, in assert_equal
[3548](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3551)
    raise error_metas[0].to_error(msg)
[3549](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3552)
AssertionError: String comparison failed: '[W OperatorEntry.cpp:159] Warning: Overri[535 chars]l)\n' != ''
[3550](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3553)
- [W OperatorEntry.cpp:159] Warning: Overriding a previously registered kernel for the same operator and the same dispatch key
[3551](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3554)
-   operator: aten::repeat_interleave.Tensor(Tensor repeats, *, int? output_size=None) -> Tensor
[3552](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3555)
-     registered at /Users/runner/work/pytorch/pytorch/build/aten/src/ATen/RegisterSchema.cpp:6
[3553](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3556)
-   dispatch key: MPS
[3554](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3557)
-   previous kernel: registered at /Users/runner/work/pytorch/pytorch/build/aten/src/ATen/RegisterCPU.cpp:30830
[3555](https://github.com/pytorch/pytorch/actions/runs/3509465096/jobs/5878922543#step:8:3558)
-        new kernel: registered at /Users/runner/work/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:48 (function registerKernel)

**Comment by changeling** [comment]
@jazzysoggy Just checking on the status of this. @kulinseth?

**Comment by changeling** [comment]
Any word?

**Comment by DenisVieriu97** [comment]
@jazzysoggy I've rebased your current branch against latest PyTorch master. Please check the results once the build is finished

**Comment by DenisVieriu97** [comment]
@jazzysoggy I've fixed the remaining failures and the kernel. Please take a look and merge it when everything is green

**Review by kulinseth — APPROVED** [review]

**Comment by changeling** [comment]
Thank y'all for all your work on this one, @jazzysoggy, @DenisVieriu97, @kulinseth! This one is much appreciated. Cheers.

**Comment by changeling** [comment]
Might be a good idea to update https://github.com/pytorch/pytorch/issues/77764 now that this has been implemented. It appears under both Good First Issue and Not categorized. :)

**Pull request closed without merging**